mqtt/mqtt-bridge/src/upstream/rpc/mod.rs

```diff
@@ -110,17 +112,56 @@ impl Display for RpcCommand {
 /// response comes back as `mqtt3::Event` type which handled with the event
 /// handler.
 #[derive(Debug, Clone, Default)]
-pub struct RpcSubscriptions(Arc<Mutex<HashMap<String, CommandId>>>);
+pub struct RpcSubscriptions(Arc<Mutex<HashMap<String, VecDeque<CommandId>>>>);
 
 impl RpcSubscriptions {
     /// Stores topic filter to command identifier mapping.
-    pub fn insert(&self, topic_filter: &str, id: CommandId) -> Option<CommandId> {
-        self.0.lock().insert(topic_filter.into(), id)
+    pub fn insert(&self, topic_filter: &str, id: CommandId) -> Option<Vec<CommandId>> {
+        let mut inner = self.0.lock();
+
+        let existing = inner
+            .get(topic_filter)
+            .map(|ids| ids.iter().cloned().collect());
+
+        inner.entry(topic_filter.into()).or_default().push_back(id);
+
+        existing
     }
 
     /// Removes topic filter to command identifier mapping and returns
     /// `CommandId` if exists.
     pub fn remove(&self, topic_filter: &str) -> Option<CommandId> {
-        self.0.lock().remove(topic_filter)
+        let mut inner = self.0.lock();
+
+        inner
+            .remove_entry(topic_filter)
+            .and_then(|(topic, mut existing)| {
+                existing.pop_front().map(|id| {
```

Comment:
Is this part necessary? Why do we need to re-insert into the hashmap? Are we not able to borrow the entry itself and then pop?

Reply:
there is a bug here :) 

answering your question: the intention here is to get a list of commands and remove collection from the hashmap if empty. my expectation that there is no more than 2 items in the collection (ideally only one). so I guessed I can remove entry and insert if needed.

Reply:
Okay okay got it 😺